src/Microsoft.OData.Client/BulkUpdateSaveResult.cs

```diff
@@ -119,13 +168,27 @@ internal void BuildDescriptorGraph<T>(IEnumerable<Descriptor> descriptors, bool
                 {
                     if (descriptor is EntityDescriptor entityDescriptor)
                     {
-                        if (entityDescriptor.ParentEntity != null && 
-                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) && 
+                        if (entityDescriptor.ParentEntity != null &&
+                            entityDescriptor.ParentEntity.Equals(topLevelDescriptor.Entity) &&
                             !bulkUpdateGraph.Contains(entityDescriptor))
                         {
                             bulkUpdateGraph.AddRelatedDescriptor(topLevelDescriptor, entityDescriptor);
                             this.BuildDescriptorGraph(descriptors, false, entityDescriptor.Entity);
                         }
+                        else
+                        {
+                            // We check if the link representation of the descriptor exists in the entity tracker and it is unchanged. 
+                            LinkDescriptor linkDescriptor = this.RequestInfo.EntityTracker.Links.Where(
+                                a => a.Source == topLevelDescriptor.Entity && 
+                                this.RequestInfo.EntityTracker.GetEntityDescriptor(a.Target) == entityDescriptor &&
+                                a.State == EntityStates.Unchanged).FirstOrDefault();

```

Comment:
@ElizabethOkerio What if we made a change to `UpdateObject` and `UpdateRelatedObject` methods such that they'd keep track of entities that are modified at any level. Then from here, all we'd need to do is map their path to the top level entity descriptors.

Reply:
@gathogojr I'm not sure I understand this. `UpdateRelatedObject` already keeps track of entities modified at whichever level. I don't think there is a clear way of `UpdateObject` knowing which parent object the object being modified belongs to. Not unless we bring the operation we are doing in the `BuildDescriptorGraph` here which will be even more expensive as it will be performed every time `updateobject` is called regardless of whether it is a `bulkupdate` or not.

Reply:
Actually, I don't think it's safe to conclude that this does not occur in a hot path. It's not uncommon for customers to make requests with expanded properties. So we shouldn't necessarily treat this as an edge case.